File: epi_judge_python/phone_number_mnemonic.py
from typing import List

from test_framework import generic_test, test_utils


def phone_mnemonic(phone_number: str) -> List[str]:
    mapping = {
        '0': ["0"],
        '1': ["1"],
        '2': ["A", "B", "C"],
        '3': ["D", "E", "F"],
        '4': ["G", "H", "I"],
        '5': ["J", "K", "L"],
        '6': ["M", "N", "O"],
        '7': ["P", "Q", "R", "S"],
        '8': ["T", "U", "V"],
        '9': ["W", "X", "Y", "Z"],
    }

    def buildm(num):
        nonlocal mapping
        if len(num) == 1:
            return mapping[num[0]]
        mnemonics = buildm(num[1:])
        res = []
        for c in mapping[num[0]]:
            for m in mnemonics:
                res.append(c + m)
        return res

    return buildm(phone_number)


# phone_mnemonic('2276696')

if __name__ == '__main__':
    exit(
        generic_test.generic_test_main(
            'phone_number_mnemonic.py',
            'phone_number_mnemonic.tsv',
            phone_mnemonic,
            comparator=test_utils.unordered_compare))

File: epi_judge_cpp/phone_number_mnemonic.cc
#include <string>
#include <vector>

#include "test_framework/generic_test.h"
using std::string;
using std::vector;

vector<string> Partial(const int index, const string &phone_number) {
  vector<string> mnemonics;
  static vector<string> num_to_alphabet{"0",   "1",   "ABC",  "DEF", "GHI",
                                        "JKL", "MNO", "PQRS", "TUV", "WXYZ"};
  for (auto &ch : num_to_alphabet[phone_number[index] - '0']) {
    if (index == 0) {
      mnemonics.push_back(string(1, ch));
    } else {
      for (auto &str : Partial(index - 1, phone_number)) {
        mnemonics.push_back(str + ch);
      }
    }
  }
  return mnemonics;
}

vector<string> PhoneMnemonic(const string &phone_number) {
  return Partial(phone_number.size() - 1, phone_number);
}

int main(int argc, char *argv[]) {
  std::vector<std::string> args{argv + 1, argv + argc};
  std::vector<std::string> param_names{"phone_number"};
  return GenericTestMain(args, "phone_number_mnemonic.cc",
                         "phone_number_mnemonic.tsv", &PhoneMnemonic,
                         UnorderedComparator{}, param_names);
}
